api/src/main/java/org/openmrs/module/oauth2/api/smart/impl/SmartAppLaunchServiceImpl.java

```diff
@@ -0,0 +1,67 @@
+package org.openmrs.module.oauth2.api.smart.impl;
+
+import java.math.BigInteger;
+import java.security.SecureRandom;
+
+import org.openmrs.api.impl.BaseOpenmrsService;
+import org.openmrs.module.oauth2.api.smart.SmartAppLaunchService;
+import org.openmrs.module.oauth2.api.smart.db.hibernate.LaunchValueDAO;
+import org.openmrs.module.oauth2.api.smart.model.LaunchValue;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.transaction.annotation.Transactional;
+
+public class SmartAppLaunchServiceImpl extends BaseOpenmrsService implements SmartAppLaunchService {
+
+	@Autowired
+	LaunchValueDAO launchValueDAO;
+
+	@Transactional
+	@Override
+	public LaunchValue saveOrUpdateLaunchValue(LaunchValue launchValue) {
+		return launchValueDAO.saveOrUpdate(launchValue);
+	}
+
+	@Transactional
+	@Override
+	public LaunchValue updateLaunchValue(LaunchValue launchValue) {
+		return launchValueDAO.update(launchValue);
+	}
+
+	@Transactional(readOnly = true)
+	@Override
+	public LaunchValue getLaunchValue(Integer smartId) {
+		return launchValueDAO.getById(smartId);
+	}
+
+	@Transactional
+	@Override
+	public LaunchValue clearLaunchValue(LaunchValue launchValue) {
+		launchValueDAO.delete(launchValue);
+		return launchValue;
+	}
+
```

Comment:
a method to get launchValue object by supplying the launchValue string..??

Reply:
Why we need such method when we already can get the LaunchValue object supplying the smartId?

Reply:
If a launchValue string fails to form a launchValue object. That means such a launchValue doesn't exist and you can discard it right away. If it exists, then you get a launchValue object and you can match anything with it next (smart_id, client_id, etc.). Nevermind, leave the optimizations for now, we'll think about them later.

Reply:
yes this seems good will do it 👍